### ISSUE TITLE
Running some tests with pkg/vm/tool/dart2 causes strong mode errors in front-end

### ISSUE THREAD

**Issue by bkonyi** [dart-lang/sdk#32101]
Came across this while debugging another issue. Looks like running some tests, in this case `hot_reload_e2e_test`, with the `pkg/vm/tool/dart2` helper script causes strong mode failures in the front-end. This test works correctly when `--strong` is removed from the helper script.

Sample output:

```
$ pkg/vm/tool/dart2 --ignore-unrecognized-flags --packages=/usr/local/google/home/bkonyi/sdk/.packages /usr/local/google/home/bkonyi/sdk/pkg/front_end/test/src/incremental/hot_reload_e2e_test.dart
pkg/front_end/lib/src/fasta/kernel/kernel_target.dart:260:14: Error: A value of type 'dart.async::Future<dynamic>' can't be assigned to a variable of type 'dart.async::FutureOr<kernel.ast::Program>'.
Try changing the type of the left hand side, or casting the right hand side to 'dart.async::FutureOr<kernel.ast::Program>'.
      return reportCrash(e, s, loader?.currentUriForCrashReporting);
             ^
pkg/front_end/lib/src/fasta/kernel/kernel_target.dart:299:14: Error: A value of type 'dart.async::Future<dynamic>' can't be assigned to a variable of type 'dart.async::FutureOr<kernel.ast::Program>'.
Try changing the type of the left hand side, or casting the right hand side to 'dart.async::FutureOr<kernel.ast::Program>'.
      return reportCrash(e, s, loader?.currentUriForCrashReporting);
             ^
pkg/front_end/lib/src/fasta/source/outline_builder.dart:736:49: Error: A value of type 'dart.core::List<dynamic>' can't be assigned to a variable of type 'dart.core::Iterable<fasta.formal_parameter_builder::FormalParameterBuilder<fasta.type_builder::TypeBuilder>>'.
Try changing the type of the left hand side, or casting the right hand side to 'dart.core::Iterable<fasta.formal_parameter_builder::FormalParameterBuilder<fasta.type_builder::TypeBuilder>>'.
        formals.setRange(count, formals.length, last);
                                                ^
pkg/front_end/lib/src/fasta/kernel/kernel_target.dart:260:14: Error: A value of type 'dart.async::Future<dynamic>' can't be assigned to a variable of type 'dart.async::FutureOr<kernel.ast::Program>'.
Try changing the type of the left hand side, or casting the right hand side to 'dart.async::FutureOr<kernel.ast::Program>'.
      return reportCrash(e, s, loader?.currentUriForCrashReporting);
             ^
pkg/front_end/lib/src/fasta/kernel/kernel_target.dart:299:14: Error: A value of type 'dart.async::Future<dynamic>' can't be assigned to a variable of type 'dart.async::FutureOr<kernel.ast::Program>'.
Try changing the type of the left hand side, or casting the right hand side to 'dart.async::FutureOr<kernel.ast::Program>'.
      return reportCrash(e, s, loader?.currentUriForCrashReporting);
             ^
pkg/front_end/lib/src/fasta/source/outline_builder.dart:736:49: Error: A value of type 'dart.core::List<dynamic>' can't be assigned to a variable of type 'dart.core::Iterable<fasta.formal_parameter_builder::FormalParameterBuilder<fasta.type_builder::TypeBuilder>>'.
Try changing the type of the left hand side, or casting the right hand side to 'dart.core::Iterable<fasta.formal_parameter_builder::FormalParameterBuilder<fasta.type_builder::TypeBuilder>>'.
        formals.setRange(count, formals.length, last);
```

**Comment by peter-ahe-google** [comment]
We should update pkg/front_end/test/fasta/bootstrap_test.dart so we catch these issues ourselves.
Also, we should probably start analyzing fasta with itself in addition to the analyzer.

**Comment by peter-ahe-google** [comment]
@stereotype441 @leafpetersen the analyzer isn't reporting these errors. Are they intentional?

**Comment by leafpetersen** [comment]
Most of these look like this bug: https://github.com/dart-lang/sdk/issues/32185 .  I believe that's (recently) fixed, maybe the SDK being used to analyze is out of date?

For this one:  pkg/front_end/lib/src/fasta/source/outline_builder.dart:736 I see a //ignore for that in the source code, which is why the analyzer is not reporting an error.

**Comment by peter-ahe-google** [comment]
Thank you, @leafpetersen It was mostly the Future<dynamic> errors I was concerned about. @bkonyi did report this 13 days ago, and  #32185 was fixed 5 days ago. This is encouraging :-)

**Comment by kmillikin** [comment]
This is a duplicate of #32185, which isn't fully fixed.  I'm closing this one and reopening that one.

**Comment by peter-ahe-google** [comment]
It's not all a duplicate:

* formals.setRange is unfortunately something I'll have to work around in Fasta.
* Also, I'll use this to track getting the strong-mode legacy bootstrapping implemented.

**Comment by peter-ahe-google** [comment]
I believe this was fixed by @bwilkerson in f66854a7008d2dc722269fd57820130c86782c66.
